simpleBalancer.Up: close of closed channel

via semaphore (https://semaphoreci.com/coreos/etcd/branches/master/builds/908)

```
=== RUN   ExampleKV_putErrorHandling
panic: close of closed channel

goroutine 615 [running]:
panic(0x1041b60, 0xc823231f80)
	/usr/local/golang/1.6.3/go/src/runtime/panic.go:481 +0x3ff
github.com/coreos/etcd/clientv3.(*simpleBalancer).Up(0xc8232008c0, 0xc8201eeb07, 0x2d, 0x0, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/clientv3/balancer.go:79 +0xa5
google.golang.org/grpc.(*addrConn).resetTransport(0xc8233e0000, 0x0, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/google.golang.org/grpc/clientconn.go:694 +0x10f6
google.golang.org/grpc.(*ClientConn).resetAddrConn.func1(0xc8233e0000)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/google.golang.org/grpc/clientconn.go:459 +0x38
created by google.golang.org/grpc.(*ClientConn).resetAddrConn
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/google.golang.org/grpc/clientconn.go:468 +0xb9
```

--- a/clientv3/balancer.go
+++ b/clientv3/balancer.go
@@ -45,6 +45,8 @@ type simpleBalancer struct {
 	// pinAddr is the currently pinned address; set to the empty string on
 	// intialization and shutdown.
 	pinAddr string
+
+	closed bool
 }
 
 func newSimpleBalancer(eps []string) *simpleBalancer {
@@ -74,15 +76,25 @@ func (b *simpleBalancer) ConnectNotify() <-chan struct{} {
 
 func (b *simpleBalancer) Up(addr grpc.Address) func(error) {
 	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	// gRPC might call Up after it called Close. We add this check
+	// to "fix" it up at application layer. Or our simplerBalancer
+	// might panic since b.upc is closed.
+	if b.closed {
+		return func(err error) {}
+	}
+
 	if len(b.upEps) == 0 {
 		// notify waiting Get()s and pin first connected address
 		close(b.upc)
 		b.pinAddr = addr.Addr
 	}
 	b.upEps[addr.Addr] = struct{}{}
-	b.mu.Unlock()
+
 	// notify client that a connection is up
 	b.readyOnce.Do(func() { close(b.readyc) })
+
 	return func(err error) {
 		b.mu.Lock()
 		delete(b.upEps, addr.Addr)
@@ -128,13 +140,19 @@ func (b *simpleBalancer) Notify() <-chan []grpc.Address { return b.notifyCh }
 
 func (b *simpleBalancer) Close() error {
 	b.mu.Lock()
+	defer b.mu.Unlock()
+	// In case gRPC calls close twice. TODO: remove the checking
+	// when we are sure that gRPC wont call close twice.
+	if b.closed {
+		return nil
+	}
+	b.closed = true
 	close(b.notifyCh)
 	// terminate all waiting Get()s
 	b.pinAddr = ""
 	if len(b.upEps) == 0 {
 		close(b.upc)
 	}
-	b.mu.Unlock()
 	return nil
 }
 

